src/laplace.jl

```diff
@@ -0,0 +1,332 @@
+    # fdot = (I - K grad2_log_p_y_given_f(f))⁻¹ Kdot grad_log_p_y_given_f(f)
+    # (I - K grad2_log_p_y_given_f(f)) = (I + K W) = (√W)⁻¹ (I + √W K √W) √W = (√W)⁻¹ B √W
+    # fdot = (√W)⁻¹ B⁻¹ √W Kdot grad_log_p_y_given_f(f)
+    ∂f_opt = cache.Wsqrt \ (cache.B_ch \ (cache.Wsqrt * (ΔK * cache.d_loglik)))
+
+    @debug "Hit frule"
+
+    return f_opt, ∂f_opt
+end
+
+function ChainRulesCore.rrule(::typeof(newton_inner_loop), dist_y_given_f, ys, K; kwargs...)
+    @debug "Hit rrule"
+    f_opt, cache = _newton_inner_loop(dist_y_given_f, ys, K; kwargs...)
+
+    # f = K (∇log p(y|f))                               (RW 3.17)
+    # δf = δK (∇log p(y|f)) + K δ(∇log p(y|f))
+    #    = δK (∇log p(y|f)) + K ∇(∇log p(y|f)) δf
+    # δf (I - K ∇∇log p(y|f)) = δK (∇log p(y|f))
+    # δf (I + K W) = δK (∇log p(y|f))
+    # δf = (I + K W)⁻¹ δK (∇log p(y|f))                 (RW 5.24)
+    # (I + K W) = (√W)⁻¹ (I + √W K √W) √W = (√W)⁻¹ B √W
+    # δf = (√W)⁻¹ B⁻¹ √W δK (∇log p(y|f))
+
+    # ∂f_opt = cache.Wsqrt \ (cache.B_ch \ (cache.Wsqrt * (ΔK * cache.d_loglik)))
+
+    # Re<Δf, δf> = Re<Δf, Wsqrt\inv B\inv Wsqrt δK d_loglik>
+    #            = Re<Wsqrt' B\inv' Wsqrt\inv' Δf d_loglik', δK>
+    #
+    # ΔK = Wsqrt' * cache.B_ch' \ Wsqrt' \ Δf_opt * cache.d_loglik'
+
+    function newton_pullback(Δf_opt)
+        ∂self = NoTangent()
+
+        ∂dist_y_given_f = @not_implemented(
+            "gradient of Newton's method w.r.t. likelihood parameters"
+        )
+
+        ∂ys = @not_implemented("gradient of Newton's method w.r.t. observations")
+
+        # ∂K = df/dK Δf
+        ∂K = @thunk(cache.Wsqrt * (cache.B_ch \ (cache.Wsqrt \ Δf_opt)) * cache.d_loglik')
+
+        return (∂self, ∂dist_y_given_f, ∂ys, ∂K)
+    end
+
+    return f_opt, newton_pullback
+end
+
+function laplace_lml(dist_y_given_f, ys, K, f_opt)
+    cache = _laplace_train_intermediates(dist_y_given_f, ys, K, f_opt)
+    return _laplace_lml(f_opt, cache)
+end
+
+function laplace_lml(
+    dist_y_given_f, ys, K; f_init=zeros(length(ys)), maxiter=100, newton_kwargs...
+)
+    f_opt = newton_inner_loop(dist_y_given_f, ys, K; f_init, maxiter, newton_kwargs...)
+    return laplace_lml(dist_y_given_f, ys, K, f_opt)
+end
+
+function laplace_lml(lfx::LatentFiniteGP, ys; newton_kwargs...)
+    dist_y_given_f, K, newton_kwargs = _check_laplace_inputs(lfx, ys; newton_kwargs...)
+    return laplace_lml(dist_y_given_f, ys, K; newton_kwargs...)
+end
+
+function laplace_f_and_lml(lfx::LatentFiniteGP, ys; newton_kwargs...)
+    dist_y_given_f, K, newton_kwargs = _check_laplace_inputs(lfx, ys; newton_kwargs...)
+    f_opt = newton_inner_loop(dist_y_given_f, ys, K; newton_kwargs...)
+    lml = laplace_lml(dist_y_given_f, ys, K, f_opt)
+    return f_opt, lml
+end
+
+function _check_laplace_inputs(
+    lfx::LatentFiniteGP, ys; f_init=nothing, maxiter=100, newton_kwargs...
+)
+    fx = lfx.fx
+    @assert mean(fx) == zero(mean(fx))  # might work with non-zero prior mean but not checked
```

Comment:
You should avoid using `@assert` and do something like `mean(fx) == zero(mean(fx)) || error("Non Zero-Mean prior not supported yet")`

Reply:
(The reason being that `@assert` can be disabled sometimes -- the docstring for `@assert` is quite informative in this regard :) )

Reply:
yeaaah that's actually an argument for keeping it as an assert :sweat_smile:

Reply:
Haha fair enough. I have no strong opinion either way!